### ISSUE TITLE
Spring Boot 3.3 upgrade causes serialization loss of DatastorePageable.urlSafeCursor

### ISSUE THREAD

**Issue by burkedavison** [GoogleCloudPlatform/spring-cloud-gcp#2940]
@odrotbohm , could you verify that the below understanding is correct? Do you have any recommendations to resolve this issue?

----

PR https://github.com/GoogleCloudPlatform/spring-cloud-gcp/pull/2923 shows the following test failure when updating from Spring Boot 3.2 to 3.3:
```
Error: 2:038 [ERROR] com.example.DatastoreBookshelfExampleIntegrationTests.testSerializedPage -- Time elapsed: 0.474 s <<< FAILURE!
java.lang.AssertionError: 

Expecting actual:
  "***"content":[***"id":12345678***],"page":***"size":1,"number":0,"totalElements":4,"totalPages":4***"
to contain pattern:
  ""pageable":.+"urlSafeCursor":".+""

	at com.example.DatastoreBookshelfExampleIntegrationTests.testSerializedPage(DatastoreBookshelfExampleIntegrationTests.java:82)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
```

[The test](https://github.com/GoogleCloudPlatform/spring-cloud-gcp/blob/main/spring-cloud-gcp-samples/spring-cloud-gcp-data-datastore-basic-sample/src/test/java/com/example/DatastoreBookshelfExampleIntegrationTests.java#L82) is verifying that `urlSafeCursor` (defined in Spring Cloud GCP's [DatastorePageable](https://github.com/GoogleCloudPlatform/spring-cloud-gcp/blob/main/spring-cloud-gcp-data-datastore/src/main/java/com/google/cloud/spring/data/datastore/repository/query/DatastorePageable.java#L30)) is part of the response body. I've tracked the change in behavior down to Spring Data's new [registration](https://github.com/spring-projects/spring-data-commons/commit/5dd7b322b65652f90717c4f8cb930c5e471ad483#diff-68a33ba2ce73d246c81458715c307eaa51a74185b682e0e09eca9730de532bfbR85-R89) of [`PagedModel`](https://github.com/spring-projects/spring-data-commons/commit/5dd7b322b65652f90717c4f8cb930c5e471ad483#diff-797fc7a0f784bf1d38dbac1f143c53bb5f8a6b2d93b637a9048e70c6dd469ab3R59-R64) which is now used when serializing a `PageImpl` with Jackson. This `PageModel` class produces the serialized `page` entry in the above 'actual' test results. This change was made "[to make sure the representations stay stable and do not expose unnecessary implementation details](https://github.com/spring-projects/spring-data-commons/commit/5dd7b322b65652f90717c4f8cb930c5e471ad483#:~:text=to%20make%20sure%20the%20representations%20stay%20stable%20and%20do%20not%20expose%20unnecessary%20implementation%20details)".

I believe the intent was to introduce this new converter as opt-in only to preserve backward compatibility, based on the [documentation for this change](https://github.com/spring-projects/spring-data-commons/commit/5dd7b322b65652f90717c4f8cb930c5e471ad483#diff-0bdc3a01c4c50e35a0bc3304ecff7bbaf6e98e91754d08cf42f7747e9195ea3bR84-R85)
> Configures how to render `PageImpl` instances. Defaults to `PageSerializationMode#DIRECT` for backward compatibility reasons.

and [the commit message](https://github.com/spring-projects/spring-data-commons/commit/5dd7b322b65652f90717c4f8cb930c5e471ad483#:~:text=with%20two%20possible%20values%3A%20DIRECT%2C%20which%20is%20the%20default%20for%20backwards%20compatibility%20reasons.)
> @EnableSpringDataWeb support now contains a pageSerializationMode attribute set to an enum with two possible values: DIRECT, which is the default for backwards compatibility reasons.

However, the logic determining the activation of this converter is only backwardly compatible for users of the `@EnableSpringDataWebSupport` annotation. If the annotation is not being used, the default behavior is to now use the `PagedModel` converter. 
```java
if (settings != null // When using the @EnableSpringDataWebSupport annotation...
    && settings.pageSerializationMode() == PageSerializationMode.DIRECT) 
{
    setMixInAnnotation(PageImpl.class, WarningMixing.class);
} else {
    // When not using the @EnableSpringDataWebSupport annotation...
    setMixInAnnotation(PageImpl.class, WrappingMixing.class);
}
// ...
@JsonSerialize(converter = PlainPageSerializationWarning.class)
abstract class WarningMixing {}

@JsonSerialize(converter = PageModelConverter.class)
abstract class WrappingMixing {} // This class uses the new PageModelConverter.
```

I believe the intent was for the logic to be:
```java
if (settings == null || settings.pageSerializationMode() == PageSerializationMode.DIRECT) {
    // Use the PlainPageSerializationWarning converter
} else {
    // Use the PageModelConverter
}
```

#### Add'l References

* Spring Data issue discussing this feature change: https://github.com/spring-projects/spring-data-commons/issues/3024
* Spring Cloud GCP sample that does not use `@EnableSpringDataWebSupport`: https://github.com/GoogleCloudPlatform/spring-cloud-gcp/tree/main/spring-cloud-gcp-samples/spring-cloud-gcp-data-datastore-basic-sample

**Comment by odrotbohm** [comment]
You're correct, and we can certainly consider this a bug. I'll have to reach out to the Boot team to find out why the current guard is not triggered. Spring Boot's [autoconfiguration for Spring Data's web support](https://github.com/spring-projects/spring-boot/blob/fa131fa6a0dbab704597ea15e3d88d763ee3d027/spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/web/SpringDataWebAutoConfiguration.java#L48) actually carries `@EnableSpringDataWebSupport`.

I've created spring-projects/spring-data-commons#3101. We should be able to get a fix for this into the next service releases (in T-2w).

**Comment by odrotbohm** [comment]
Fixes should be available in both the 3.3.1 and 3.4.0 snapshots.

**Comment by burkedavison** [comment]
Thank you for the fast resolution, @odrotbohm !
